### PR TITLE
[Linux] GDB JIT support for source line debug information.

### PR DESCRIPTION
**Pull request by Dmitri-Botcharnikov**
This PR adds generation of source line debug information according to GDB JIT interface for using with LLDB (see #6011).
The feature is enabled by providing its own flag to cmake, i.e. `./build.sh cmakeargs -DFEATURE_GDBJIT=TRUE`.  It also requires the latest SymbolReader managed dll to work (dotnet/corefx#9571).
The JIT notification is generated at the same time as the DAC notification in `DACNotify::DoJITNotification`, so when managed breakpoint is requested by `bpmd` command, the source lines are displayed on break when available.
```
(lldb) bpmd Test.exe Test.Main
Adding pending breakpoints...
(lldb) c
Process 14998 resuming
1 location added to breakpoint 1
(lldb) JITTED Test!Test.Main()
Setting breakpoint: breakpoint set --address 0xAFB56461 [Test.Main()]
Process 14998 stopped
* thread #1: tid = 14998, 0xafb56460 JIT(0xbc230) at Test.cs:87, name = 'corerun', stop reason = breakpoint 3.1
    frame #0: 0xafb56460 JIT(0xbc230) at Test.cs:87
   84  
   85  
   86       static int Main()
-> 87       {
   88           DumpIL();
   89           LikelyInlined();
   90           UnlikelyInlined();

```
The current implementation requires source files in current directory and portable PDB in original or current directory.
Tested in Linux/x64 and Linux/ARM environments.

### CODE REVIEW

File: src/vm/CMakeLists.txt [hunk 1]
```diff
@@ -27,6 +27,17 @@ if(CLR_CMAKE_PLATFORM_UNIX)
     add_compile_options(-fPIC)
 endif(CLR_CMAKE_PLATFORM_UNIX)
 
+if(FEATURE_GDBJIT)
+    set(VM_SOURCES_GDBJIT
+    ../gc/gccommon.cpp
```
Comment (swgillespie) [inline src/vm/CMakeLists.txt]: Why are these necessary?

File: src/vm/gdbjit.cpp [hunk 2]
```diff
@@ -0,0 +1,1075 @@
...
+        delete[] dbgAbbrev.MemPtr;
+        delete[] dbgLine.MemPtr;
+        delete[] dbgStr.MemPtr;
+        return;
+    }
+    
+    /* Build .debug_pubname section */
+    if (!BuildDebugPub(dbgPubname, methodName, dbgInfo.MemSize, 26))
+    {
+        delete[] symInfo;
```
Comment (janvorli) [inline src/vm/gdbjit.cpp]: I recommend using NewArrayHolder wrapper for all of these arrays. Then you don't have to do any cleanup explicitly and repeat it over and over, the destructor of the NewArrayHolder would do that for you. When / if you need to keep a pointer over the lifetime of the holder, you just call SuppressRelease on the holder.

File: src/vm/gdbjit.h [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+
+/* Type for section indices, which are 16-bit quantities.  */
+typedef uint16_t Elf32_Section;
+typedef uint16_t Elf64_Section;
+
+/* The ELF file header.  This appears at the start of every ELF file.  */
+
+#define EI_NIDENT (16)
+
+struct  Elf32_Ehdr
```
Comment (jkotas) [inline src/vm/gdbjit.h]: These definitions looks like copy & paste from glibc (https://github.com/bminor/glibc/blob/master/elf/elf.h). 

https://github.com/dotnet/coreclr/blob/master/Documentation/project-docs/contributing.md for rules that need to be followed for reusing code from another projects.

File: src/coreclr/hosts/inc/coreclrhost.h [hunk 4]
```diff
@@ -46,5 +46,22 @@ CORECLR_HOSTING_API(coreclr_execute_assembly,
             unsigned int* exitCode);
 
 #undef CORECLR_HOSTING_API
-                      
+
+struct SequencePointInfo
+{
+    int lineNumber, ilOffset;
+    char16_t* fileName;
```
Comment (janvorli) [inline src/coreclr/hosts/inc/coreclrhost.h]: The hosting API is 8 bit char based. Can we have a char here too?

File: src/coreclr/hosts/inc/coreclrhost.h [hunk 5]
```diff
@@ -46,5 +46,22 @@ CORECLR_HOSTING_API(coreclr_execute_assembly,
...
+};
+
+struct MethodDebugInfo
+{
+    SequencePointInfo* points;
+    int size;
+};
+
+typedef int (*GetInfoForMethodDelegate)(const char*, unsigned int, MethodDebugInfo& methodDebugInfo);
+extern GetInfoForMethodDelegate getInfoForMethodDelegate;
```
Comment (janvorli) [inline src/coreclr/hosts/inc/coreclrhost.h]: We should add an API to set the delegate instead of having the variable in the hosting API.

Comment (janvorli) [inline src/coreclr/hosts/inc/coreclrhost.h]: Actually, it is kind of strange to create the delegate in the coreruncommon.cpp, which would mean that anyone implementing another hosting app would have to add it there too. It seems it would better be part of the coreclr_initialize, completely hidden inside coreclr.

Comment (lucenticus) [inline src/coreclr/hosts/inc/coreclrhost.h]: I've added commit with creation of coreclr delegate inside coreclr_initialize. 
@janvorli, please take a look whether I understood your comment correctly.

File: src/coreclr/hosts/inc/coreclrhost.h (hunk 4 above)
Comment (lucenticus) [inline src/coreclr/hosts/inc/coreclrhost.h]: `fileName` is marshaled to managed code as string in [SymbolReader](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.Debug.SymbolReader/src/System/Diagnostics/Debug/SymbolReader/SymbolReader.cs#L23). `CharSet` can be either `Auto, Ansi, or Unicode`. Is it possible to setup `CharSet` to UTF-8? Perhaps I misunderstood your comment?

File: src/coreclr/hosts/inc/coreclrhost.h [hunk 6]
```diff
@@ -46,5 +46,7 @@ CORECLR_HOSTING_API(coreclr_execute_assembly,
             unsigned int* exitCode);
 
 #undef CORECLR_HOSTING_API
-                      
+
```
Comment (janvorli) [inline src/coreclr/hosts/inc/coreclrhost.h]: Can you please remove the whitespace changes from this file?

File: src/coreclr/hosts/unixcoreruncommon/CMakeLists.txt [hunk 7]
```diff
@@ -2,6 +2,10 @@ project(unixcoreruncommon)
 
 add_compile_options(-fPIC)
 
+if(FEATURE_GDBJIT)
```
Comment (janvorli) [inline src/coreclr/hosts/unixcoreruncommon/CMakeLists.txt]: Now that you have moved the stuff to the hosting api, this is not needed here anymore

File: src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp [hunk 8]
```diff
@@ -30,7 +30,6 @@
 // If set to 1, server GC is enabled on startup. If 0, server GC is
 // disabled. Server GC is off by default.
 static const char* serverGcVar = "CORECLR_SERVER_GC";
-
```
Comment (janvorli) [inline src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp]: Can you please remove the whitespace change here?

File: src/dlls/mscoree/unixinterface.cpp [hunk 9]
```diff
@@ -238,8 +243,21 @@ int coreclr_initialize(
...
+        hr = coreclr_create_delegate(*hostHandle,
+                                     *domainId,
+                                     "System.Diagnostics.Debug.SymbolReader",
+                                     "System.Diagnostics.Debug.SymbolReader.SymbolReader",
+                                     "GetInfoForMethod",
+                                     (void**)&getInfoForMethodDelegate);
+
+        if (!SUCCEEDED(hr))
+        {
+            fprintf(stderr, "coreclr_create_delegate failed - status: 0x%08x\n", hr);
```
Comment (janvorli) [inline src/dlls/mscoree/unixinterface.cpp]: Could you please change the error message to be more specific? For example by adding the full name of the function (`System.Diagnostics.Debug.SymbolReader.SymbolReader.GetInfoForMethod`) to the message.

File: src/vm/CMakeLists.txt (hunk 1 above)
Comment (janvorli) [inline src/vm/CMakeLists.txt]: I also wonder why you need this change. The GC is now a separate static library.

File: src/inc/utilcode.h [hunk 10]
```diff
@@ -5750,4 +5750,21 @@ extern SpinConstants g_SpinConstants;
 
 // ======================================================================================
 
+#ifdef FEATURE_GDBJIT
```
Comment (janvorli) [inline src/inc/utilcode.h]: I would prefer creating a separate header src/vm/gdbjit.h instead of sticking this stuff into the utilcode.h.

File: src/coreclr/hosts/inc/coreclrhost.h (hunk 4 above)
Comment (janvorli) [inline src/coreclr/hosts/inc/coreclrhost.h]: Now that you have moved the stuff inside of the hosting api code, you can ignore this comment.

File: src/vm/CMakeLists.txt [hunk 11]
```diff
@@ -27,6 +27,13 @@ if(CLR_CMAKE_PLATFORM_UNIX)
     add_compile_options(-fPIC)
 endif(CLR_CMAKE_PLATFORM_UNIX)
 
+if(FEATURE_GDBJIT)
+    set(VM_SOURCES_GDBJIT
+    gdbjit.cpp
+    )
+   add_definitions(-DFEATURE_GDBJIT)
```
Comment (janvorli) [inline src/vm/CMakeLists.txt]: A nit - wrong indentation (3 spaces instead of 4)

File: src/vm/CMakeLists.txt [hunk 12]
```diff
@@ -27,6 +27,13 @@ if(CLR_CMAKE_PLATFORM_UNIX)
     add_compile_options(-fPIC)
 endif(CLR_CMAKE_PLATFORM_UNIX)
 
+if(FEATURE_GDBJIT)
+    set(VM_SOURCES_GDBJIT
+    gdbjit.cpp
```
Comment (janvorli) [inline src/vm/CMakeLists.txt]: Can you please indent this line?

**Comment by manu-silicon** [comment]
@Dmitri-Botcharnikov You mention that the source has to be in the current working directory. Is this restriction easy to lift?

**Comment by Dmitri-Botcharnikov** [comment]
@manu-silicon Not so easy. Probably we need to generate search directory table in DWARF .debug_line section, but we didn't investigated this.

File: src/vm/gdbjit.h (hunk 3 above)
Comment (lucenticus) [inline src/vm/gdbjit.h]: Whether appropriate to use code from LLVM headers (`ELF.h `and `DWARF.h`)?

Comment (jkotas) [inline src/vm/gdbjit.h]: I think so. You should still follow the above link. It may be best to add the two LLVM headers in separate preparatory PR.

File: src/vm/gdbjit.cpp [hunk 13]
```diff
@@ -314,75 +311,42 @@ void NotifyGdb::MethodCompiled(MethodDesc* MethodDescPtr)
...
-        delete[] dbgPubname.MemPtr;
-        delete[] dbgPubType.MemPtr;
-        delete[] sectStr.MemPtr;
         return;
     }
 
     /* Patch section offsets & sizes */
     long offset = sizeof(Elf_Ehdr);
-    Elf_Shdr* pShdr = reinterpret_cast<Elf_Shdr*>(sectHeaders.MemPtr);
+    Elf_Shdr* pShdr = reinterpret_cast<Elf_Shdr*>(sectHeaders.MemPtr.Extract());
```
Comment (janvorli) [inline src/vm/gdbjit.cpp]: It seems you really want GetValue instead of Extract (and the same at all other places). The Extract method suppresses release of the pointer, so the memory would not be freed by the destructor, while the GetValue just gets the value of the pointer.

Comment (lucenticus) [inline src/vm/gdbjit.cpp]: I've fixed that, thanks for clarification!

File: src/vm/gdbjit.h (hunk 3 above)
Comment (lucenticus) [inline src/vm/gdbjit.h]: I've created PR: https://github.com/dotnet/coreclr/pull/6383

**Comment by lucenticus** [comment]
@janvorli We've fixed all your comments, PTAL. Is it ready to merge?

File: src/dlls/mscoree/unixinterface.cpp [hunk 14]
```diff
@@ -238,8 +246,24 @@ int coreclr_initialize(
...
+                                     "GetInfoForMethod",
+                                     (void**)&getInfoForMethodDelegate);
+
+        if (!SUCCEEDED(hr))
+        {
+            fprintf(stderr,
+                    "Can't create delegate for 'System.Diagnostics.Debug.SymbolReader.SymbolReader.GetInfoForMethod' "
+                    "method - status: 0x%08x\n", hr);
+        }
+        return S_OK; // We don't need to fail if we can't create delegate
```
Comment (janvorli) [inline src/dlls/mscoree/unixinterface.cpp]: Can you just set hr to S_OK instead of returning it please?

**Comment by janvorli** [comment]
LGTM modulo my last comment.

**Comment by janvorli** [comment]
@lucenticus Thank you!

File: src/dlls/mscoree/unixinterface.cpp (hunk 14 above)
Comment (lucenticus) [inline src/dlls/mscoree/unixinterface.cpp]: Done, thanks!

**Comment by mmitche** [comment]
@dotnet-bot test this please

**Comment by lucenticus** [comment]
@dotnet-bot test  Windows_NT x86 ryujit Checked Build and Test  please

**Comment by jkotas** [comment]
cc @mikem8361 @noahfalk

**Comment by chunseoklee** [comment]
@lucenticus 
On my x64 ubuntu, there is build break on cross-build to arm32 : 
```
/home/twoflower/dev/coreclr/src/vm/gdbjit.cpp:272:74: error: no member named
      'GetGCInfoToken' in 'EECodeInfo'
  ...codeSize = codeInfo.GetCodeManager()->GetFunctionSize(codeInfo.GetGCInfo...
                                                           ~~~~~~~~ ^
1 error generated.
make[2]: *** [src/vm/wks/CMakeFiles/cee_wks.dir/__/gdbjit.cpp.o] Error 1
make[1]: *** [src/vm/wks/CMakeFiles/cee_wks.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
/home/twoflower/dev/coreclr/src/vm/gdbjit.cpp:272:74: error: no member named
      'GetGCInfoToken' in 'EECodeInfo'
  ...codeSize = codeInfo.GetCodeManager()->GetFunctionSize(codeInfo.GetGCInfo...
                                                           ~~~~~~~~ ^
1 error generated.
```

**Comment by lucenticus** [comment]
@dotnet-bot test this please

**Comment by lucenticus** [comment]
@chunseoklee Hm, I've tried arm32 clean cross build and it was build without any errors. Did you rebase to the latest master?

**Comment by chunseoklee** [comment]
@lucenticus For internal reason, I rebased on 32a020e47a0797bba4167e85efc38a34397f633e commit. Of course, your two commits related to ELF.h are applied.

**Comment by lucenticus** [comment]
@chunseoklee I've changed `GetGCInfo` -> `GetGCInfoToken` because there are some changes in the latest master branch. If you checkout to older master, then you have to change back  `GetGCInfoToken` -> `GetGCInfo`

**Comment by lucenticus** [comment]
@dotnet-bot test  Linux ARM Emulator Cross Debug Build please
@dotnet-bot test Linux ARM Emulator Cross Release Build please

**Comment by Dmitri-Botcharnikov** [comment]
@dotnet-bot test CentOS7.1 x64 Debug Build and Test please

**Comment by Dmitri-Botcharnikov** [comment]
@dotnet-bot test Windows_NT x86 ryujit Checked Build and Test please

**Comment by noahfalk** [comment]
Thanks for putting this together :+1: 

A few questions:
a) What happens if the user uses bpmd in the code implementing the symbol reader (or anything in its call graph)
b) Am I correct to understand this only generates source mapping for functions where bpmd was used to place a breakpoint. If a user sets a breakpoint in one function but then steps to another function the 2nd function does not have line mappings? (Not saying it has to, but hoping to understand the current limitations)
c) Do you know how many total bytes of data gets generated for each method? If it is suitably large we might want to have some environment variable that turns this behavior on/off.

**Comment by Dmitri-Botcharnikov** [comment]
@noahfalk 
a) Now it leads to crash. We don't know the legal way to automatically separate user code from system.
b) Yes, you're right. We've experimented with generation of debug info for each jitted method but it leads to the same problem as in previous item. It would be great if we had some facility to detect for which methods the debug info is to be generated.
c) The generated data for method with several lines  is about 1 KB on x64 and about 700 bytes  on arm32 (ELF64 vs ELF32). The GDB/JIT interface supports jitted code dropping where we can free memory, but as far as I know this feature is not enabled in current version CoreCLR. We also wrapped the implementation with FEATURE_GDBJIT macro so you can just build without it.

**Comment by noahfalk** [comment]
FEATURE_GDBJIT is enabled by default in this PR right? If this feature creates a potentially large and hard to calculate set of system methods that will cause the debuggee to crash when you use bpmd on them I think we'll need a better mitigation than asking developers to rebuild the runtime with custom defines. 

In order of preference I'd suggest:
1) changing the implementation so it doesn't crash (even if it fails to give debug data for every usage of bpmd)
2) warning the user at the time they used bpmd that the method they chose is problematic and they need to take some additional action to prevent the runtime from crashing. This would need to be something that is reasonable to do in a debugger without rebuilding the runtime and ideally without restarting the app.
3) make the feature opt-in instead of opt-out 

For 1 does a reentrancy check suffice to prevent the crash? (ie set a thread local flag before calling into calling the symbol reader delegate and disable generating any debug data while that flag is set). 

I'm not sure 2 has any easy implementation, but someone else might see options I am missing.

For 3 perhaps developers can opt-in to feature at debug time by changing the value of a runtime global flag using the debugger? That keeps the feature pretty easy to turn on when you want it.

**Comment by jkotas** [comment]
> FEATURE_GDBJIT is enabled by default in this PR right?

It is not enabled by default. It is enabled by providing its own flag to cmake per the PR description at the top.

**Comment by Dmitri-Botcharnikov** [comment]
@jkotas is right,  the feature is not enabled by default by this PR. 
Concerning global flag. What if the feature would be enabled by environment variable that lists names of modules for which debugging information should be generated? I.e:
```
export CORECLR_GDBJIT='myapp.exe,mylibrary.dll'
```
Consequently, if the environment variable is unset then the feature is disabled.
In this case we can also drop the compile time flag.
Implementing a reentrancy preventing flag is also great idea. Could you please point which API we should use for that?

**Comment by noahfalk** [comment]
Ah sorry, I missed that in the description entirely. If it is not enabled by default then I have no concern.

As for the re-entrancy check I'm not sure it would work, but what I had in mind was something simple like this:

    __declspec(thread) bool tls_isSymReaderInProgress = false; // probably need to do this in a more compiler agnostic way

    void DACNotify::DoJITNotification(MethodDesc *MethodDescPtr)
    {
        if(!tls_isSymReaderInProgress)
        {
             tls_isSymReaderInProgress = true;
             NotifyGdb::MethodCompiled(MethodDescPtr);
             tls_isSymReaderInProgress = false;
        }

       ...

**Comment by lucenticus** [comment]
@dotnet-bot test  Linux ARM Emulator Cross Debug Build please

**Comment by noahfalk** [comment]
I see you added the reentrancy check, does that mean it worked to avoid the crash?

**Comment by mikem8361** [comment]
Not that I know a lot about the gdb JIT support, but LGTM.

**Comment by lucenticus** [comment]
@noahfalk Yes, I implemented reentrancy check, it doesn't crash now while trying to debug SymbolReader itself. We continue work on adding environment variable with a list of debugged modules, but I think it will be a separate PR. This PR is ready to merge in case you don't have any comments.

**Comment by noahfalk** [comment]
Nope I'm good, in we go : )

**Comment by lucenticus** [comment]
@noahfalk Thanks!
